src/net_processing.cpp

```diff
@@ -3234,6 +3270,7 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                     }
                     state->m_tx_download.m_tx_in_flight.erase(in_flight_it);
                     state->m_tx_download.m_tx_announced.erase(inv.hash);
```

Comment:
Now that we're potentially rerequesting txs from different peers after receiving a NOTFOUND, I think we _shouldn't_ erase them from a peer's _m_tx_in_flight_ and _m_tx_announced_ structures in receipt of a `NOTFOUND`, and keep them there until the `TX_EXPIRY_INTERVAL` in `SendMessages()`, otherwise there might be an attack where an adversary has multiple connections to you, and then juggles you between them by sending an `INV` for a transaction, holding on to the `GETDATA`, and then sending a `NOTFOUND` immediately followed by an `INV` reannouncing the same tx.

Reply:
Hmm, that means a peer that sends a NOTFOUND is treated exactly the same as one that just ignores the request; so NOTFOUND is just a "fyi, I'm never going to respond, so feel free to try someone else". That... seems like it makes a lot of sense? If we were to ever start punishing peers for (frequently) not responding to tx requests, maybe it would make sense to punish them for (equally frequently) sending notfounds too?